src/sinks/gcp/mod.rs

```diff
@@ -69,45 +69,53 @@ pub struct GcpCredentials {
 }
 
 fn get_token_implicit() -> Result<Token, GcpError> {
-    let mut rt = tokio_compat::runtime::current_thread::Runtime::new().unwrap();
-    rt.block_on_std(async {
-        let req = http::Request::get(SERVICE_ACCOUNT_TOKEN_URL)
-            .header("Metadata-Flavor", "Google")
-            .body(hyper::Body::empty())
-            .unwrap();
-
-        let res = hyper::Client::new()
-            .request(req)
-            .await
-            .context(GetImplicitToken)?;
-
-        let body = res.into_body();
-        let bytes = hyper::body::to_bytes(body).await.context(GetTokenBytes)?;
-
-        // Token::from_str is irresponsible and may panic!
-        match serde_json::from_slice::<Token>(&bytes) {
-            Ok(token) => Ok(token),
-            Err(error) => Err(match serde_json::from_slice::<TokenErr>(&bytes) {
-                Ok(error) => GcpError::TokenFromJson { source: error },
-                Err(_) => GcpError::TokenJsonFromStr { source: error },
-            }),
-        }
+    std::thread::spawn(|| {
+        let mut rt = crate::runtime::Runtime::single_threaded().unwrap();
+        rt.block_on_std(async {
+            let req = http::Request::get(SERVICE_ACCOUNT_TOKEN_URL)
+                .header("Metadata-Flavor", "Google")
+                .body(hyper::Body::empty())
+                .unwrap();
+
+            let res = hyper::Client::new()
+                .request(req)
+                .await
+                .context(GetImplicitToken)?;
+
+            let body = res.into_body();
+            let bytes = hyper::body::to_bytes(body).await.context(GetTokenBytes)?;
+
+            // Token::from_str is irresponsible and may panic!
+            match serde_json::from_slice::<Token>(&bytes) {
+                Ok(token) => Ok(token),
+                Err(error) => Err(match serde_json::from_slice::<TokenErr>(&bytes) {
+                    Ok(error) => GcpError::TokenFromJson { source: error },
+                    Err(_) => GcpError::TokenJsonFromStr { source: error },
+                }),
+            }
+        })
     })
+    .join()
+    .expect("Runtime thread error")
 }
 
 fn get_token_with_creds_blocking(
-    jwt: &Jwt<JwtClaims>,
-    credentials: &Credentials,
+    jwt: Jwt<JwtClaims>,
+    credentials: Credentials,
 ) -> Result<Token, GoErr> {
-    let mut rt = tokio_compat::runtime::current_thread::Runtime::new()?;
-    rt.block_on_std(goauth::get_token(jwt, credentials))
+    std::thread::spawn(move || {
+        let mut rt = crate::runtime::Runtime::single_threaded().unwrap();
+        rt.block_on_std(async move { goauth::get_token(&jwt, &credentials).await })
+    })
```

Comment:
I don't understand the need to create a thread here (and above) to block on single threaded. Could the async nature be pushed up a couple levels and block on elsewhere? If not, it should probably be documented here why we're going to this work.

Reply:
If we call `block_on_std` from new runtime in already working runtime we get error about tokio timer. This happens because tokio timer use thread local variable: https://docs.rs/tokio-timer/0.3.0-alpha.5/src/tokio_timer/timer/handle.rs.html#77
Creating thread creates new thread local variable and this allows call `block_on_std` from another runtime.
We can push async up on a few levels, but `Sink::build` still will need creating runtime.. this should be solved soon.

Reply:
I added a comment and moved async on few levels up